### PR TITLE
SW-5626 Implement ApplicationStore

### PR DESCRIPTION
**Pull request by sgrimm**
Implement the business logic for most of the operations on applications required
to support pre-screening.

This does not include any notification hooks; those will come later since they're
not needed for pre-screening.

### CODE REVIEW

**Comment by sgrimm** [comment]
* **#2166** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2165** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2165?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2164** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2163** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2163?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2162** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2162?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sgrimm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+  fun fetchOneById(applicationId: ApplicationId): ExistingApplicationModel {
+    requirePermissions { readApplication(applicationId) }
+
+    return fetchByCondition(APPLICATIONS.ID.eq(applicationId)).firstOrNull()
+        ?: throw ApplicationNotFoundException(applicationId)
+  }
+
+  fun fetchByProjectId(projectId: ProjectId): List<ExistingApplicationModel> {
+    requirePermissions { readProject(projectId) }
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt:42]: This allows org members below admin to readApplication, which is more relaxed than `readApplication` I think

File: src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+    return fetchByCondition(APPLICATIONS.PROJECT_ID.eq(projectId))
+  }
+
+  fun fetchByOrganizationId(organizationId: OrganizationId): List<ExistingApplicationModel> {
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt:47]: same here

File: src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+          val internalName = "${alpha3CountryCode}_$organizationName"
+
+          updateInternalName(applicationId, internalName)
+        } else {
+          log.debug(
+              "Not setting internal name for application $applicationId because boundary is not " +
+                  "all in one country: $countries")
+        }
+      }
+
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt:176]: should we validate the boundary size constraint here as well?

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt:176]: I just posted a question on the PRD about this. The size checking behavior is a little inconsistent as specified.

File: src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt (hunk 1 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt:42]: Good catch. Added code to filter the results.

**Review by tommylau523 — APPROVED** [review]

**Comment by sgrimm** [comment]
### Merge activity

* **Jul 17, 2:41 PM EDT**: @sgrimm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2165).
* **Jul 17, 3:00 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2165) rebased this pull request as part of a merge.
* **Jul 17, 3:08 PM EDT**: @sgrimm merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2165).

**Pull request merged**